packages/starlight/__e2e__/test-utils.ts

```diff
@@ -1,34 +1,79 @@
 // provide a Starlight Playwright fixture accessible from within all tests.
-export async function testFactory(fixturePath: string) {
-	let previewServer: PreviewServer | undefined;
+export function testFactory(fixturePath: string) {
+	const fixturePathUrl = new URL(fixturePath, import.meta.url);
+	// Combining absolute paths with a `file:` base URL on Windows results
+	// in a URL with the drive letter as the protocol.
+	// In that case, use the URL as is instead of interpreting the `file:` protocol.
+	const root =
+		fixturePathUrl.protocol === 'file:'
+			? fileURLToPath(new URL(fixturePath, import.meta.url))
+			: fixturePathUrl.toString();
 
-	const test = baseTest.extend<{ starlight: StarlightPage }>({
-		starlight: async ({ page }, use) => {
-			if (!previewServer) {
-				throw new Error('Could not find a preview server to run tests against.');
-			}
+	async function makeServer(
+		options: {
+			mode?: 'build' | 'dev';
+		} = {}
+	): Promise<Server> {
+		const { mode } = options;
+		if (mode === 'dev') {
+			return await dev({ logLevel: 'error', root });
+		} else {
+			await build({ logLevel: 'error', root });
+			return await preview({ logLevel: 'error', root });
+		}
+	}
 
-			await use(new StarlightPage(previewServer, page));
-		},
-	});
+	// Optimization for tests that don't customize any server options
+	// to not rebuild the fixture for each test.
+	let prodServer: Server | null = null;
+	const servers = new Map<string, Server>();
 
-	test.beforeAll(async () => {
-		const root = fileURLToPath(new URL(fixturePath, import.meta.url));
-		await build({ logLevel: 'error', root });
-		previewServer = await preview({ logLevel: 'error', root });
+	const test = baseTest.extend<{
+		getProdServer: () => Promise<StarlightPage>;
+		makeServer: (name: string, ...params: Parameters<typeof makeServer>) => Promise<StarlightPage>;
+	}>({
+		getProdServer: ({ page }, use) =>
+			use(async () => {
+				const server = (prodServer ??= await makeServer({
+					mode: 'build',
+				}));
+				return new StarlightPage(server, page);
+			}),
+		makeServer: ({ page }, use) =>
+			use(async (name, ...params) => {
+				const server = servers.get(name) ?? (await makeServer(...params));
+				servers.set(name, server);
+				return new StarlightPage(server, page);
+			}),
 	});
 
-	test.afterAll(async () => {
-		await previewServer?.stop();
+	test.afterAll(({}, context) => {
+		// Playwright's afterAll timeout is shared with the last test
+		// in the suite. If the last test is slower, stopping all the
+		// servers can easily read the timeout limit.
+		// To avoid that, we add 40 seconds to effective timeout limit.
+		context.setTimeout(context.timeout + 40000);
+		return Promise.race([
+			// Let the cleanup happen on the background
+			timers.setTimeout(30000),
+			(async () => {
+				// Stop all started servers.
+				await prodServer?.stop();
+				for (const server of servers.values()) {
+					await server.stop();
+				}
+			})(),
+		]);
```

Comment:
I think the Playwright behavior described in the comment makes sense but I think something may be wrong here.

At this point, the last test likely succeeded and we only have to cleanup resources. `context.setTimeout(context.timeout + 40000);` would add 40 seconds to the timeout just to cleanup resources. And after that, the `Promise.race` is basically saying, you have 30 seconds to cleanup resources otherwise I'll continue.

I think the code I would expect here would be:

```ts
await prodServer?.stop();
for (const server of servers.values()) {
  await server.stop();
}
```

and nothing else.

If I clone this branch, change the `test.afterAll` code to the one above, edit `packages/starlight/__e2e__/git.test.ts` to transform the last `test` (`include git information while developing`) to a `test.only` and run the E2E tests, this single test would fail due a timeout issue.
If I add a bit of logging, I can see that the last (and single) expectation of this test is met in 2s on my machine so this means that the test is taking more than 38s to cleanup resources.
If I add back at the beggining of the `test.afterAll` the line `context.setTimeout(context.timeout + 40000);`, and even use a ridiculous high value, the test still times out.
If I edit the `packages/starlight/playwright.config.ts` file and switch `headless` to `false`, the test passes in 2s altho an `✘ [ERROR] The build was canceled` message is logged.

I think something doesn't make sense here. I'm not even sure what this "build" message is about as we should not be building anything by running only this specific test. But it definitely looks like closing a dev server is hanging the test indefinitely.

I'm assuming this problem and the current approach (with the `Promise.race`) may also explain why the time to run the tests before and after these changes almost doubled).

I'll try to investigate this further but I don't think I've seen anything obvious so far that could be causing this issue.

Reply:
I see this problem only on CI. Localy it passes in about 5s even when stopping one server at a time and waiting.

Reply:
Wow, that's even more weird. For me, this single test definitely hangs at `await server.stop();`. I definitely wonder what could be the cause of such weird/inconsistent behavior.

Reply:
Adding a link to a Discord message containing the investigation I did on this topic so I don't lose it: https://discord.com/channels/830184174198718474/1070481941863878697/1271515653840568341